### PR TITLE
[Java Client] Fix wrong behavior of deduplication for key based batching

### PR DESCRIPTION
**Pull request by BewareMyPower**
### Motivation

Currently message deduplication doesn't work well for key based
batching. First, the key based batch container doesn't update the
`lastSequenceIdPushed`. So a batch could contain both duplicated and not
duplicated messages. Second, when `createOpSendMsgs` is called, the
`OpSendMsg` objects are sorted by the lowest sequence ids, and the
highest sequence id is not set. If a batch contains sequence id 0,1,2,
then the message with sequence id 1 or 2 won't be dropped.

### Modifications

- Refactor the key based batch container that the
  `BatchMessageContainerImpl` is reused instead of maintaining a
  `KeyedBatch` class.
- When `createOpSendMsgs` is called, clear the highest sequence id field
  and configure the sequence id field with the highest sequence id to fix
  the second issue described before.
- Add `testKeyBasedBatchingOrder` to show and verify the current
  behavior.
- Add test for key based batching into
  `testProducerDeduplicationWithDiscontinuousSequenceId` to verify
  `lastSlastSequenceIdPushed` is updated correctly.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Comment by BewareMyPower** [comment]
Though this PR changed the current behavior that the `OpSendMsg`s are **now sorted by the highest sequence id instead of the lowest sequence id**. I think it should be a bug fix rather than a feature. So I added the labels for older branches. If anyone worried about this behavior change, feel free to point it out.

**Comment by poorbarcode** [comment]
Hi @BewareMyPower. I don't think we've solved the problem. 

- The original implementation, you think there might be this duplication: 

```
// batch 1th
send [seq_1, seq_3]
// batch 2th
send [seq_2, seq_3]
```
So the message which is sequence_3 was repeated. 

- After this piece of code changes, is there such a repeated message ?
```
// batch 1th
send [seq_1, seq_5]
// batch 2th
send [seq_1, seq_3]
```

So the message which is sequence_1 was repeated.  


I don't know whether I have correctly understood your meaning. If there is any misunderstanding, thank you for telling me.

**Comment by BewareMyPower** [comment]
First, both cases should not be valid. User should be responsible to make the sequence id unique.

What this PR wants to solve is the **resend** case of an existing message. For example, 4 messages might be split into two groups in key based batching:

- A: 0, 3 (i.e. messages whose keys are all "A", and the sequence ids are 0 and 3)
- B: 1, 2

If the client didn't receive the ack receipt of batch A, it might resend batch A (msg-0 and msg-3). There are two possible cases:
1. batch A has been persisted -> broker should not accept the resend
2. batch A has not been persisted -> broker should accept the resend 

Assuming msg-0 and msg-3 are sent **individually (without batch)**.

For the 1st case, we can verify it easily with the unit test.

```java
    public void test() throws Exception {
        final String topic = "persistent://my-property/my-ns/test";
        admin.namespaces().setDeduplicationStatus("my-property/my-ns", true);

        final Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
                .topic(topic)
                .batcherBuilder(BatcherBuilder.KEY_BASED)
                .create();
        final List<TypedMessageBuilder<String>> messagesToSend = new ArrayList<>();
        messagesToSend.add(producer.newMessage().value("msg-0").key("A").sequenceId(0));
        messagesToSend.add(producer.newMessage().value("msg-1").key("B").sequenceId(1));
        messagesToSend.add(producer.newMessage().value("msg-2").key("B").sequenceId(2));
        messagesToSend.add(producer.newMessage().value("msg-3").key("A").sequenceId(3));
        final List<CompletableFuture<MessageId>> futures = new ArrayList<>();
        messagesToSend.forEach(msg -> futures.add(msg.sendAsync()));
        producer.flush(); // flush the messages in the batch container
        final List<MessageId> ids = futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
        // Resend messages in batch A
        ids.add(messagesToSend.get(0).send());
        ids.add(messagesToSend.get(3).send());
        // TODO: print the ids
    }
```

|       | Before this patch | After this patch |
| ----- | ----------------- | ---------------- |
| msg-0 | rejected          | rejected         |
| msg-3 | **accepted**      | rejected         |

Before this patch, the resend of `msg-3` is accepted, this behavior must be wrong. Because `msg-0` and `msg-3` have already been persisted.

For the 2nd case, we can modify the test to:

```java
        final List<CompletableFuture<MessageId>> futures = new ArrayList<>();
        // Assume msg-0 and msg-3 are lost
        futures.add(messagesToSend.get(1).sendAsync());
        futures.add(messagesToSend.get(2).sendAsync());
        producer.flush(); // flush the messages in the batch container
```


|       | Before this patch | After this patch |
| ----- | ----------------- | ---------------- |
| msg-0 | rejected          | rejected         |
| msg-3 | accepted          | accepted         |

The behaviors are the same. Though before this patch, `msg-3` is accepted because `3 > 1`, while after this patch, `msg-3` is accepted because `3 > 2`.

**It's a pity to see `msg-0` is rejected as the duplicated message.** It's true that this PR cannot solve this problem as well.

However, if we resent `msg-0` and `msg-3` as a batch,

```java
        // Resend messages in batch A
        futures.add(messagesToSend.get(0).sendAsync());
        futures.add(messagesToSend.get(3).sendAsync());
        final List<MessageId> newIds = futures.stream().map(CompletableFuture::join).collect(Collectors.toList());
```

|       | Before this patch | After this patch    |
| ----- | ----------------- | ------------------- |
| msg-0 | **rejected**      | **Never Completed** |
| msg-3 | **rejected**      | **Never Completed** |

> After this patch, the futures of `sendAsync` never completed. It might be another bug to fix.
>
> > 2022-05-03T04:17:20,988+0800 [main] WARN  org.apache.pulsar.client.impl.ProducerImpl - Message with sequence id 0 is definitely a duplicate

----

In short, this PR solves the problem that duplicated messages might be accepted even if the previous messages were persisted successfully.

- With the default batching builder, the highest sequence id is updated in broker.
- (Before this patch) With the key based batching builder, the lowest sequence id is updated in broker.

**Comment by BewareMyPower** [comment]
To make it more clear and simple, https://github.com/apache/pulsar/pull/5491 introduced the highest sequence id to `MessageMetadata` so that broker could update the latest sequence id to the sequence id of **the sequence id of the latest message in the batch**. However, the current implementation of the key based batching never tells the broker **the sequence id of the latest message in the batch**.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java [hunk 1]
```diff
@@ -115,66 +82,46 @@ public boolean isMultiBatches() {
...
+            // In key based batching, the sequence ids might not be ordered, for example,
+            // | key | sequence id list |
+            // | :-- | :--------------- |
+            // | A | 0, 3, 4 |
+            // | B | 1, 2 |
+            // The message order should be 1, 2, 0, 3, 4 so that a message with a sequence id <= 4 should be dropped.
+            // However, for a MessageMetadata with both `sequence_id` and `highest_sequence_id` fields, the broker will
+            // expect a strict order so that the batch of key "A" (0, 3, 4) will be dropped.
+            // Therefore, we should update the `sequence_id` field to the highest sequence id and remove the
+            // `highest_sequence_id` field to allow the weak order.
```
Comment (BewareMyPower) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: > It's a pity to see msg-0 is rejected as the duplicated message. It's true that this PR cannot solve this problem as well.

In addition to [my comment](https://github.com/apache/pulsar/pull/15413#issuecomment-1115325307), it's an issue of the sequence ids cannot be strictly ordered in key based batching. They could only be ordered per key.

Therefore, here we noted that key based batching allows the sequence id order like 1, 2, 0, 3, 4 if 0, 3, 4 are messages of another key. @poorbarcode

Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: > What this PR wants to solve is the **resend** case of an existing message. For example, 4 messages might be split into two groups in key based batching:
> 
> * A: 0, 3 (i.e. messages whose keys are all "A", and the sequence ids are 0 and 3)
> * B: 1, 2
> For the 1st case, we can verify it easily with the unit test.

Thank you for your explanation, I see what you mean.  I still have a question:

```java
public void test() throws Exception {
        final String topic = "persistent://my-property/my-ns/test";
        admin.namespaces().setDeduplicationStatus("my-property/my-ns", true);

        final Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
                .topic(topic)
                .batcherBuilder(BatcherBuilder.KEY_BASED)
                .create();
        final List<TypedMessageBuilder<String>> messagesToSend = new ArrayList<>();
        messagesToSend.add(producer.newMessage().value("msg-0").key("A").sequenceId(0));
        messagesToSend.add(producer.newMessage().value("msg-1").key("B").sequenceId(1));
        messagesToSend.add(producer.newMessage().value("msg-3").key("A").sequenceId(3));
        final List<CompletableFuture<MessageId>> futures = new ArrayList<>();
        messagesToSend.forEach(msg -> futures.add(msg.sendAsync()));
        producer.flush(); 
        // didn't receive the ack receipt of batch B,  it might resend batch B.
        ids.add(messagesToSend.get(1).send());
 }
```
If the client didn't receive the ack receipt of batch B,  it might resend batch B.

-- | Before this patch | After this patch
-- | -- | --
msg-1 | accepted | rejected

This patch changed this implements.  

Anyway, I think it's an good catch !

**Review by poorbarcode — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java (hunk 1 above)
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: > If the client didn't receive the ack receipt of batch B, it might resend batch B.

After this patch the send order will be batch B, batch A as they are sorted by the highest sequence id (1 < 3).

If batch A was not persisted, it would be OK. Otherwise, since batch B was sent before batch A, it means batch A has been persisted as well. So rejecting batch B is reasonable.

Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: > After this patch the send order will be batch B, batch A as they are sorted by the highest sequence id (1 < 3).

If batch B has [msg-1, msg-4], then client didn't receive the ack receipt of batch B, it might resend batch B. like this

```
ids.add(messagesToSend.get(1).send());
ids.add(messagesToSend.get(4).send());
```

after this patch, msg-1 will be rejected. msg-4 will be accepted.

Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: > msg-1 will be rejected

**It's a correct behavior.** This behavior could be incorrect only when **batch A has been persisted while batch B is not persisted**.  However, batch B must be persisted before batch A.

Comment (poorbarcode) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]:
```java
batches.values().forEach(batchMessageContainer -> {
                batchMessageContainer.setLowestSequenceId(batchMessageContainer.getHighestSequenceId());
                batchMessageContainer.setHighestSequenceId(-1L);
            });
            return batches.values().stream().sorted((o1, o2) ->
                    (int) (o1.getLowestSequenceId() - o2.getLowestSequenceId())
            )
```

If batch A has [msg-0, msg-3] and batch B has [msg-1, msg-4],  batch B will persistent after batch A. 

So problems can still happen

Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:97]: Yeah. It's right as I've already explained.

**Comment by codelipenghui** [comment]
```java
messagesToSend.add(producer.newMessage().value("msg-0").key("A").sequenceId(0));
messagesToSend.add(producer.newMessage().value("msg-1").key("B").sequenceId(1));
messagesToSend.add(producer.newMessage().value("msg-2").key("B").sequenceId(2));
messagesToSend.add(producer.newMessage().value("msg-3").key("A").sequenceId(3));
```

What is the expected behavior for the above case? 

The deduplication depends on the monotonically increasing sequence ID, but the key-based
batcher will break the rule.

**Comment by BewareMyPower** [comment]
> ```java
> messagesToSend.add(producer.newMessage().value("msg-0").key("A").sequenceId(0));
> messagesToSend.add(producer.newMessage().value("msg-1").key("B").sequenceId(1));
> messagesToSend.add(producer.newMessage().value("msg-2").key("B").sequenceId(2));
> messagesToSend.add(producer.newMessage().value("msg-3").key("A").sequenceId(3));
> ```
> 
> What is the expected behavior for the above case?
> 
> The deduplication depends on the monotonically increasing sequence ID, but the key-based batcher will break the rule.

|                            | Before this patch  | After this patch   |
| -------------------------- | ------------------ | ------------------ |
| message order              | A-0, A-3, B-1, B-2 | B-1, B-2, A-0, A-3 |
| highest sequence id pushed | 1                  | 3                  |

I've explained in details in the previous comment. It's true that even after this patch, message deduplication cannot achieve effectively-once, instead, it's at-most once. But it can avoid the **duplicated messages**. It doesn't solve the problem fundamentally, but it make it better.

I noted that when #4435 added the key based batch container, it also sorted the batches by sequence id.

https://github.com/apache/pulsar/blob/a1fb200ff707e9855efb563a27a894664a59c58b/pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java#L154-L156

At that time, #5491 was not pushed and there is no highest sequence id field in `MessageMetadata`. In #5491, it looks like the key based batch container was forgotten so that there were no changes.

Even without considering the enhancement (maybe meaningless at this time) to the deduplication, the refactor of this PR that reuses the `BatchMessageContainerImpl` rather than the `KeyedBatch` could make the code easier to maintain. For example, when `currentTxnidMostBits` and `currentTxnidLeastBits` were added in #8415, these two fields must be added to both default and key based containers.

BTW, maybe in future, we can support deduplication by keys and `Key_Shared` mode can also support effectively-once as well.

**Comment by BewareMyPower** [comment]
The initial purpose of this PR was that I'm wondering why adding a `KeyedBatch` class in `BatchMessageKeyBasedContainer` rather than reusing the `BatchMessageContainerImpl`. Then I found the difference that  `BatchMessageKeyBasedContainer` 
1. doesn't update the `ProducerImpl#lastSequenceIdPushed`
2. sort the batches by sequence id, not highest sequence id

So during the refactoring, I tend to adopt the more correct behavior rather than keeping the previous behavior (might not be on purpose, but a mistake).

**Comment by codelipenghui** [comment]
Oh, I see. I thought the PR was for support deduplication for key-based batching before.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ClientDeduplicationTest.java [hunk 2]
```diff
@@ -46,7 +64,7 @@ protected void cleanup() throws Exception {
         super.internalCleanup();
     }
 
-    @Test
+    @Test(priority = -1)
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/api/ClientDeduplicationTest.java:67]: Any reason for setting priority to `-1`?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java [hunk 3]
```diff
@@ -115,66 +82,46 @@ public boolean isMultiBatches() {
...
+            batches.values().forEach(batchMessageContainer -> {
+                batchMessageContainer.setLowestSequenceId(batchMessageContainer.getHighestSequenceId());
+                batchMessageContainer.setHighestSequenceId(-1L);
```
Comment (codelipenghui) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:100]: We can remove this line? we have set the sequence ID with the `highestSequenceId`, we will get a same behavior between `highestSequenceId == -1` or `highestSequenceId = sequenceId`

**Review by codelipenghui — COMMENTED** [review]
The change looks good to me, just left some minor comments.

File: pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java [hunk 4]
```diff
@@ -519,6 +519,8 @@ public static ByteBufPair newSend(long producerId, long sequenceId, long highest
                 .setSequenceId(sequenceId);
         if (highestSequenceId >= 0) {
             send.setHighestSequenceId(highestSequenceId);
+        } else {
+            send.clearHighestSequenceId();
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-common/src/main/java/org/apache/pulsar/common/protocol/Commands.java:523]: Yeah, we don't need to clear it here.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java [hunk 5]
```diff
@@ -169,6 +187,10 @@ public void discard(Exception ex) {
             if (firstCallback != null) {
                 firstCallback.sendComplete(ex);
             }
+            if (batchedMessageMetadataAndPayload != null) {
+                ReferenceCountUtil.safeRelease(batchedMessageMetadataAndPayload);
+                batchedMessageMetadataAndPayload = null;
+            }
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java:193]: Yes.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java [hunk 6]
```diff
@@ -49,37 +38,21 @@
...
-            if (msg.getMessageBuilder().hasTxnidLeastBits() && currentTxnidLeastBits == -1) {
-                currentTxnidLeastBits = msg.getMessageBuilder().getTxnidLeastBits();
-            }
-
-        } else {
-            part.addMsg(msg, callback);
+        final BatchMessageContainerImpl batchMessageContainer = batches.computeIfAbsent(key,
+                __ -> new BatchMessageContainerImpl(producer));
+        batchMessageContainer.add(msg, callback);
+        if (batchMessageContainer.isAdded()) {
```
[the remark this answers is not in the document]
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:53]: Yes.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/api/ClientDeduplicationTest.java (hunk 2 above)
Comment (BewareMyPower) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/api/ClientDeduplicationTest.java:67]: We need to run `testNamespaceDeduplicationApi` before all other tests because it assumes the initial duplication status is null.

```java
    public void testNamespaceDeduplicationApi() throws Exception {
        final String namespace = "my-property/my-ns";
        assertNull(admin.namespaces().getDeduplicationStatus(namespace));
```

After `testKeyBasedBatchingOrder` is added, `testNamespaceDeduplicationApi` will run after `testKeyBasedBatchingOrder`.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java (hunk 3 above)
Comment (BewareMyPower) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageKeyBasedContainer.java:100]: Yes.

**Comment by BewareMyPower** [comment]
@codelipenghui All comments are addressed, PTAL again.

**Review by codelipenghui — APPROVED** [review]

**Comment by BewareMyPower** [comment]
@hangc0276 Could you also take a look?

**Comment by BewareMyPower** [comment]
I accidentally found (triggered) a bug of GitHub that I merged a PR twice because of the network issue.:sweat_smile:

image

This PR is still in an Open state, However, please don't merge this PR again.
